src/explore-education-statistics-common/src/modules/table-tool/utils/applyTableHeadersOrder.ts

```diff
@@ -0,0 +1,439 @@
+import { TableHeadersConfig } from '@common/modules/table-tool/types/tableHeaders';
+import {
+  CategoryFilter,
+  LocationFilter,
+  TimePeriodFilter,
+  Filter,
+} from '@common/modules/table-tool/types/filters';
+import isEqual from 'lodash/isEqual';
+import intersectionWith from 'lodash/intersectionWith';
+import intersection from 'lodash/intersection';
+
+type GroupsKeys = 'rowGroups' | 'columnGroups';
+type LastGroupKeys = 'rows' | 'columns';
+type Axes = 'row' | 'column';
+
+interface Keys {
+  groupsKey: GroupsKeys;
+  lastGroupKey: LastGroupKeys;
+  otherAxisGroupsKey: GroupsKeys;
+  otherAxisLastGroupKey: LastGroupKeys;
+}
+const getKeys = (axis: Axes): Keys => {
+  return {
+    groupsKey: `${axis}Groups`,
+    lastGroupKey: `${axis}s`,
+    otherAxisGroupsKey: axis === 'row' ? 'columnGroups' : 'rowGroups',
+    otherAxisLastGroupKey: axis === 'row' ? 'columns' : 'rows',
+  };
+};
+
+const getFilterType = (filter: Filter): string => {
+  if (filter instanceof CategoryFilter) {
+    return (filter as CategoryFilter).category;
+  }
+  if (filter instanceof LocationFilter) {
+    return 'locationFilter';
+  }
+  if (filter instanceof TimePeriodFilter) {
+    return 'timePeriodFilter';
+  }
+  return 'indicator';
+};
+
+const getMatchingGroup = (
+  groups: Filter[][],
+  groupToMatch: Filter[],
+): Filter[] | undefined => {
+  const groupToMatchFilterType = getFilterType(groupToMatch[0]);
+  return groups.find(
+    group => getFilterType(group[0]) === groupToMatchFilterType,
+  );
+};
+
+/**
+ * Order items within a group based on the previous reordering.
+ * Items that weren't in the reordered list will be after the reordered ones.
+ */
+const orderItems = (
+  reorderedItems: Filter[],
+  updatedItems: Filter[],
+): Filter[] => {
+  return reorderedItems
+    .filter(item =>
+      updatedItems.find(updatedItem => updatedItem.value === item.value),
+    )
+    .concat(
+      updatedItems.filter(
+        updatedItem =>
+          !reorderedItems.find(item => item.value === updatedItem.value),
+      ),
+    );
+};
+
+const isReorderedGroup = (
+  originalGroup: Filter[],
+  updatedGroup: Filter[],
+): boolean => {
+  const intersectionOrderedByOriginal = intersectionWith(
+    originalGroup,
+    updatedGroup,
+    isEqual,
+  );
+  const intersectionOrderedByUpdated = intersectionWith(
+    updatedGroup,
+    originalGroup,
+    isEqual,
+  );
+  return !isEqual(intersectionOrderedByOriginal, intersectionOrderedByUpdated);
+};
+
+/**
+ * Have the groups in an axis been re-ordered or moved between axes
+ * */
+const areGroupsReordered = (
+  previousOrder: string[],
+  previousOtherAxisOrder: string[],
+  updatedOrder: string[],
+  updatedOtherAxisOrder: string[],
```

Comment:
Would use object syntax to make the parameters more explicit. Are they're all the same type, it'd be really easy to put one of the parameters in the wrong position and it wouldn't be noticed.

```ts
function areGroupReordered({
  previousOrder,
  previousOtherAxisOrder,
  updatedOrder,
  updatedOtherAxisOrder
} : {
  previousOrder: string[],
  previousOtherAxisOrder: string[],
  updatedOrder: string[],
  updatedOtherAxisOrder: string[],
}): boolean {
```

Don't have to, but could add the return type of `boolean` too to make things nice and explicit what this function does. We could also do this in our various other functions as well.